### DISCUSSION TITLE
Close drawer/overlay on back button instead of full refresh?

### DISCUSSION THREAD

**Discussion by jogaman** [unpoly/unpoly#434] [General]
Hi,
Is there a way I can get unpoly to not do a full refresh on the back button, and instead close the drawer/overlay?

Say, I go from /one/ to /one/two/ to /one/two/three/ and then press back, I'll get a server side rendered version of /one/two/, thus destroying the existing stack. It would be great if the topmost drawer was simply closed (same as pressing X) instead.

possible? Thanks!

**Comment by adam12** [comment]
Can you enable logging and then paste the logs into here? `up.log.enable()`. I thought this was already the functionality but maybe I'm misremembering.

**Comment by jogaman** [comment]
Yes! Here it is:
I go from / to /first-overlay/, then a second drawer /someurl/second-overlay/.
Then the problem is when i press the back button and I get a server side rendered version of /first-overlay/

```html

Invalid URI. Load of media resource  failed. theurl.com
 __ _____  ___  ___  / /_ __
/ // / _ \/ _ \/ _ \/ / // /  2.6.1
\___/_//_/ .__/\___/_/\_. / 
        / /            / /

Load unpoly-migrate.js to enable deprecated APIs.

Call `up.log.disable()` to disable logging for this session. unpoly.min.js:1:117147
up.hello() Compiling fragment  
<body up-source="/">
unpoly.min.js:1:116604
up:fragment:inserted Inserted fragment  
<body up-source="/">
unpoly.min.js:1:116604
up:link:follow Following link unpoly.min.js:1:116604
up:request:load Loading GET /first-overlay/ unpoly.min.js:1:116604
up.request() Change with { solo } option will abort other requests unpoly.min.js:1:116604
up:request:late Server is slow to respond unpoly.min.js:1:116604
up:request:loaded Server responded HTTP 200 to GET /first-overlay/ (31790 characters) unpoly.min.js:1:116604
up:request:recover Slow requests were loaded unpoly.min.js:1:116604
up:fragment:loaded Loaded fragment from successful response to GET /first-overlay/ unpoly.min.js:1:116604
up.render() Could not match primary target (.base). Updating a fallback target (main). unpoly.min.js:1:116604
up.render() Opening element "main" in new overlay unpoly.min.js:1:116604
up:layer:open Opening new overlay unpoly.min.js:1:116604
up:location:changed Advanced to location /first-overlay/ unpoly.min.js:1:116604
up.hello() Compiling fragment  
<up-drawer position="right" size="full" nesting="0">
unpoly.min.js:1:116604
up:fragment:inserted Inserted fragment  
<up-drawer position="right" size="full" nesting="0">
unpoly.min.js:1:116604
up:layer:opened Opened new drawer overlay unpoly.min.js:1:116604
up:link:follow Following link unpoly.min.js:1:116604
up:request:load Loading GET /someurl/second-overlay/ unpoly.min.js:1:116604
up.request() Change with { solo } option will abort other requests unpoly.min.js:1:116604
up:request:loaded Server responded HTTP 200 to GET /someurl/second-overlay/ (11781 characters) unpoly.min.js:1:116604
up:fragment:loaded Loaded fragment from successful response to GET /someurl/second-overlay/ unpoly.min.js:1:116604
up.render() Could not match primary target ([up-main~=drawer]). Updating a fallback target (main). unpoly.min.js:1:116604
up.render() Opening element "main" in new overlay unpoly.min.js:1:116604
up:layer:open Opening new overlay unpoly.min.js:1:116604
up:location:changed Advanced to location /someurl/second-overlay/ unpoly.min.js:1:116604
up.hello() Compiling fragment  
<up-drawer position="left" size="medium" nesting="1">
unpoly.min.js:1:116604
up:fragment:inserted Inserted fragment  
<up-drawer position="left" size="medium" nesting="1">
unpoly.min.js:1:116604
up:layer:opened Opened new drawer overlay unpoly.min.js:1:116604
GEThttp://theurl.com/favicon.ico
[HTTP/1.1 404 Not Found 22ms]

up:request:load Loading GET /first-overlay/ unpoly.min.js:1:116604
up:request:loaded Server responded HTTP 200 to GET /first-overlay/ (31790 characters) unpoly.min.js:1:116604
up:fragment:loaded Loaded fragment from successful response to GET /first-overlay/ unpoly.min.js:1:116604
up.render() Updating "body" in root layer unpoly.min.js:1:116604
up:layer:dismiss Will dismiss drawer overlay with value  :peel unpoly.min.js:1:116604
up:layer:dismissed Dismissed drawer overlay with value  :peel unpoly.min.js:1:116604
up:layer:dismiss Will dismiss drawer overlay with value  :peel unpoly.min.js:1:116604
up:location:changed Advanced to location / unpoly.min.js:1:116604
up:layer:dismissed Dismissed drawer overlay with value  :peel unpoly.min.js:1:116604
up:location:changed Advanced to location /first-overlay/ unpoly.min.js:1:116604
up.hello() Compiling fragment  
<body up-source="/first-overlay/">
unpoly.min.js:1:116604
up:fragment:inserted Inserted fragment  
<body up-source="/first-overlay/">
unpoly.min.js:1:116604
up.viewport.restoreScroll() Restoring scroll positions for URL /first-overlay/ to  
Object {  }
unpoly.min.js:1:116604
up:location:changed Restored location /first-overlay/ unpoly.min.js:1:116604
Invalid URI. Load of media resource  failed. first-overlay

​


```

**Comment by jogaman** [comment]
Looking at the line that say
`up:location:changed Advanced to location / unpoly.min.js:1:116604`

..could it be related to this issue I encountered before? https://github.com/unpoly/unpoly/issues/397

**Reply by adam12** [reply]
It's entirely possible.

What backend are you using Unpoly with? Can you make a reproduction using our [Glitch template](https://glitch.com/edit/#!/empty-unpoly-project?path=index.html%3A1%3A0)?

**Comment by jogaman** [comment]
Here is a glitch https://ultra-fifth-dresser.glitch.me/

Using Processwire - so probably close to regular PHP?

**Comment by adam12** [comment]
Thanks for the Glitch example. I must be mistaken about functionality. I tested the demo site, and when hitting the back button a few layers in, it re-renders the last layer outside of the modal/drawer. I'm assuming that's the correct functionality. Perhaps Henning remembers otherwise.

1. Visit https://demo.unpoly.com/projects
2. Click New project. See modal.
3. Click New company. See new layer modal stacked.
4. Press Back button. See New project page rendered full screen.

**Comment by jogaman** [comment]
Hello,
Thank you. Yes, I suppose this is how it's designed to work. I have an audio player that breaks because of the page reload. So because of this, I was wondering if there is a way I can work around it? I'm considering

1. Listen to popstate
2. If underlying layer is the same url as previous url then peel() topmost layer
3. cancel the event somehow and prevent unpoly to do anything except peel of the top most layer

If you have any idea if this wouldn't work or would work, I'd be curious to know. Otherwise I'll give it a shot. But I have a feeling a hack like this will lead to some odd behaviors :P

**Comment by triskweline** [comment]
> I tested the demo site, and when hitting the back button a few layers in, it re-renders the last layer outside of the modal/drawer. I'm assuming that's the correct functionality.

This is the designed behavior. Navigating through history will open the history entry's URL as a full page.

> Say, I go from /one/ to /one/two/ to /one/two/three/ and then press back, I'll get a server side rendered version of /one/two/, thus destroying the existing stack. It would be great if the topmost drawer was simply closed (same as pressing X) instead.

I can understand where this idea is coming from, but it's non-trivial to implement this consistently. Your quoted example is the really the simplest case. The previous URL could also have been produced by *closing* an overlay. So going back would need to *restore* it. Also remember that you need to support forward navigation as well.

> I have an audio player that breaks because of the page reload

Maybe you can move the audio element outside the layer stack (e.g. the end of the `body`). To make it survive history navigation you can either use `[up-keep]` or configure [`up.history.config.restoreTargets`](https://unpoly.com/up.history.config#config.restoreTargets) to something more specific than `'body'`, like a layout container element.

**Reply by jogaman** [reply]
> I can understand where this idea is coming from, but it's non-trivial to implement this consistently. Your quoted example is the really the simplest case. The previous URL could also have been produced by _closing_ an overlay. So going back would need to _restore_ it. Also remember that you need to support forward navigation as well.

I understand. I suppose all layers underneath would have to be reloaded. Too costly of an operation, I suppose.

> Maybe you can move the audio element outside the layer stack (e.g. the end of the `body`). To make it survive history navigation you can either use `[up-keep]` or configure [`up.history.config.restoreTargets`](https://unpoly.com/up.history.config#config.restoreTargets) to something more specific than `'body'`, like a layout container element.

Fantastic, had no idea about restoreTargets. Made a quick test and it works great!

Thanks for taking the time with this.
